src/model_repository_manager.h

```diff
@@ -72,12 +73,73 @@ class ModelRepositoryManager {
     const std::string reason_;
   };
 
+  // Information about the model.
+  struct ModelInfo {
+    ModelInfo(
+        const int64_t mtime_nsec, const int64_t prev_mtime_ns,
+        const std::string& model_path)
+        : mtime_nsec_(mtime_nsec), prev_mtime_ns_(prev_mtime_ns),
+          explicitly_load_(true), model_path_(model_path),
+          is_config_provided_(false)
+    {
+    }
+    ModelInfo()
+        : mtime_nsec_(0), prev_mtime_ns_(0), explicitly_load_(true),
+          is_config_provided_(false)
+    {
+    }
+    int64_t mtime_nsec_;
+    int64_t prev_mtime_ns_;
+    bool explicitly_load_;
+    inference::ModelConfig model_config_;
+    std::string model_path_;
+    // Temporary location to hold agent model list before creating the model
+    // the ownership must transfer to ModelLifeCycle to ensure
+    // the agent model life cycle is handled properly.
+    std::shared_ptr<TritonRepoAgentModelList> agent_model_list_;
+    bool is_config_provided_;
+  };
+
+  // Map from model name to information about the model.
+  struct ModelInfoMap {
+    ModelInfoMap() = default;
+    ModelInfoMap(const ModelInfoMap& rhs);
+    ModelInfoMap& operator=(const ModelInfoMap& rhs);
+
+    void Swap(ModelInfoMap& rhs);
+    Status GetModelInfo(
+        const ModelIdentifier& model_id, ModelInfo** model_info) const;
+
+    // Write updated model info back to this object after model load/unload.
+    void Writeback(
+        const ModelInfoMap& updated_model_info,
+        const std::set<ModelIdentifier>& affected_models);
+
+    std::unordered_map<ModelIdentifier, std::unique_ptr<ModelInfo>> map_;
+  };
+
+  // std::condition_variable wrapper with mutex included
+  class ConditionVariable {
+   public:
+    ConditionVariable() = default;
+    ConditionVariable(const ConditionVariable&) = delete;
+    ConditionVariable& operator=(const ConditionVariable&) = delete;
+
+    void wait();
+    void notify_all();
+
+   private:
+    std::mutex mu_;
+    std::condition_variable cv_;
+  };
+
   /// A basic unit in dependency graph that records the models seen by the model
   /// repository manager.
   struct DependencyNode {
     DependencyNode(const ModelIdentifier& model_id)
         : status_(Status::Success), model_id_(model_id), checked_(false),
-          connected_(false)
+          connected_(false), is_locked_(false),
+          conflict_retry_cv_(new ConditionVariable())
```

Comment:
Each node holds one of this class at initialization?

Reply:
Yes, because the load/unload conflict could be caused by any node (model), but at any time, there will only be one of this class per model, which is why it is held by a shared pointer to be shared between copies of the node.